### PR TITLE
 add liftering support to MFCC inversion 

### LINKED ISSUE #1006: MFCC inversion should support liftering
#### Description
#940 implemented liftering in the MFCC extractor, but we didn't add inverse-liftering to the inverse-mfcc function.

This would be relatively easy to do, as inverse-liftering would divide out the lifter (raised sine) before applying the inverse-DCT transform using this calculation: https://github.com/librosa/librosa/blob/master/librosa/feature/spectral.py#L1697

Note that for some settings of `lifter`, this could result in divide-by-zero errors, and it should raise a warning in that case.

### PR DESCRIPTION
**Pull request by SuperKogito**
#### Reference Issue

fixes #1006 

#### What does this implement/fix? Explain your changes.
I added a new input variable called `inv_lifter` to [mfcc_to_mel](https://github.com/librosa/librosa/blob/master/librosa/feature/inverse.py#167)  and [mfcc_to_audio](https://github.com/librosa/librosa/blob/master/librosa/feature/inverse.py#217), which is used in a similar fashion to `lifter` in  [mfcc](https://github.com/librosa/librosa/blob/master/librosa/feature/spectral.py#1571).
- if `inv_lifter` is positive, then inverse the liftering by mfcc division.
- if `inv_lifter` is zero, then no inverse liftering is performed.
- if `inv_lifter` is negative, then raise a `ParameterError`.

The docs are adjusted accordingly.

### CODE REVIEW

**Comment by bmcfee** [comment]
Thanks for getting this started!  It generally looks good, but I have a few comments:

1. Can you change the parameter name from `inv_lifter` to `lifter` so that it matches with the forward transformation?
2. Your implementation applies inverse-liftering in-place (`mfcc /= ...`) which will modify the input data because it's passed by reference.  Please change it to operate on a copy instead.
3. Can you add test coverage for the new functionality in `mfcc_to_mel`?  I don't think we need a full forward-backward comparison, but at least check the following cases:
  a. warnings are triggered when appropriate
  b. the lifter parameter is used properly -- this can be tested with, say, an all-ones mfcc vector, so that `idct(ones * lifter)` can be pre-computed for a couple of different lifter values.

**Comment by SuperKogito** [comment]
I was not sure if you meant a shallow or deep copy so I just followed what is used in other librosa scripts. There was also an error in the previous implementation causing a size mismatch that I now solved .

Point 3.b. is still not covered.
I need first to find some pre-defined cases as you mentioned so I can test if the lifter is used properly. 
I will try to get it done over the weekend.

File: tests/test_features.py [hunk 1]
```diff
@@ -958,10 +958,21 @@ def test_mfcc_to_mel(n_mfcc, n_mels, dct_type):
                                             dct_type=dct_type,
                                             lifter=10**-3)
 
-    # check if mfcc_to_mel it works correctly with lifter
-    librosa.feature.inverse.mfcc_to_mel(mfcc, n_mels=n_mels,
-                                        dct_type=dct_type,
-                                        lifter=1)
+    # check if mfcc_to_mel works correctly with lifter
+    ones = np.ones(mfcc.shape)
+    negate_lifter_vec = 1 + (lifter / 2) * np.sin(np.pi * np.arange(1, 1 +
```
Comment (lostanlen) [inline tests/test_features.py]: this nonsyntactic line break is difficult to parse for the human reader. it breaks at the worst possible place in the equation, at the sixth level of nesting.

Comment (SuperKogito) [reply tests/test_features.py]: the line breaks are automatically done using [autopep8](https://github.com/hhatto/autopep8). Although, this is practical but I admit that it can result in some weird line breaks. [yapf](https://github.com/jpadilla/yapf-online) seems to provide better results. Shortening the variable name and using yapf results in: 
```
    lifter_vec = 1 + (lifter / 2) * np.sin(np.pi * np.arange(
        1, 1 + n_mfcc, dtype=mfcc.dtype) / lifter)[:, np.newaxis]
```
If you think this variant is better, I will edit the code accordingly. Otherwise, I suggest a no line break at all in this position.

Comment (lostanlen) [reply tests/test_features.py]: you can also assign a temporary variable for `np.arange(1, 1 + n_mfcc, dtype=mfcc.dtype)`

File: librosa/feature/inverse.py [hunk 2]
```diff
@@ -195,26 +196,44 @@ def mfcc_to_mel(mfcc, n_mels=128, dct_type=2, norm='ortho', ref=1.0):
     ref : number or callable
         Reference power for (inverse) decibel calculation
 
+    lifter : number >= 0
+        If `lifter>0`, apply *inverse-liftering* (inverse cepstral filtering):
```
Comment (lostanlen) [inline librosa/feature/inverse.py]: `*inverse-liftering*` -> `inverse liftering`
(no need for `*` nor hyphen)

File: librosa/feature/inverse.py [hunk 3]
```diff
@@ -195,26 +196,44 @@ def mfcc_to_mel(mfcc, n_mels=128, dct_type=2, norm='ortho', ref=1.0):
...
     Returns
     -------
     M : np.ndarray [shape=(n_mels, n)]
         An approximate Mel power spectrum recovered from `mfcc`
 
+    Warns
+    --------
+    RuntimeWarning
+        overflow encountered in power or
+        divide by zero encountered during inverse-liftering.
```
Comment (lostanlen) [inline librosa/feature/inverse.py]: inverse liftering (no hyphen)

File: librosa/feature/inverse.py [hunk 4]
```diff
@@ -195,26 +196,44 @@ def mfcc_to_mel(mfcc, n_mels=128, dct_type=2, norm='ortho', ref=1.0):
...
     --------
     mfcc
     melspectrogram
     scipy.fftpack.dct
     '''
+    if lifter > 0:
+        mfcc = mfcc.copy()
+        n_mfcc = mfcc.shape[0]
+        mfcc /= 1 + (lifter / 2) \
+            * np.sin(np.pi * np.arange(1, 1 + n_mfcc, dtype=mfcc.dtype) / lifter)[:, np.newaxis]
```
Comment (lostanlen) [inline librosa/feature/inverse.py]: declare a different variable `liftered_mfcc`, so that there's no need need for `.copy()`

Comment (lostanlen) [reply librosa/feature/inverse.py]: also define the constant `np.sin(np.pi * np.arange(1, 1 + n_mfcc, dtype=mfcc.dtype) / lifter)` on a separate line, with an informative name (`lifter_sine` ?)

File: librosa/feature/inverse.py [hunk 5]
```diff
@@ -195,26 +196,44 @@ def mfcc_to_mel(mfcc, n_mels=128, dct_type=2, norm='ortho', ref=1.0):
...
 
-    logmel = scipy.fftpack.idct(mfcc, axis=0, type=dct_type, norm=norm, n=n_mels)
+        # avoid division of zero by zero errors
+        mfcc[np.isnan(mfcc)] = 1.
+
+    elif lifter != 0:
+        raise ParameterError('MFCC to MEL lifter={} must be a non-negative number'.format(lifter))
```
Comment (lostanlen) [inline librosa/feature/inverse.py]: MEL -> mel

File: librosa/feature/inverse.py [hunk 6]
```diff
@@ -195,26 +196,44 @@ def mfcc_to_mel(mfcc, n_mels=128, dct_type=2, norm='ortho', ref=1.0):
...
     '''
+    if lifter > 0:
+        mfcc = mfcc.copy()
+        n_mfcc = mfcc.shape[0]
+        mfcc /= 1 + (lifter / 2) \
+            * np.sin(np.pi * np.arange(1, 1 + n_mfcc, dtype=mfcc.dtype) / lifter)[:, np.newaxis]
 
-    logmel = scipy.fftpack.idct(mfcc, axis=0, type=dct_type, norm=norm, n=n_mels)
+        # avoid division of zero by zero errors
+        mfcc[np.isnan(mfcc)] = 1.
```
Comment (lostanlen) [inline librosa/feature/inverse.py]: is this really needed? can't you catch this before applying the division?
i would prefer to avoid multiple sequential assignments on the same array. they tend to be error-prone. ideally, the output is computed in a single pass.

File: librosa/feature/inverse.py [hunk 7]
```diff
@@ -244,6 +263,10 @@ def mfcc_to_audio(mfcc, n_mels=128, dct_type=2, norm='ortho', ref=1.0, **kwargs)
     ref : number or callable
         Reference power for (inverse) decibel calculation
 
+    lifter : number >= 0
+        If `lifter>0`, apply *inverse-liftering* (inverse cepstral filtering):
```
Comment (lostanlen) [inline librosa/feature/inverse.py]: again, no asterisks, no hyphen please

**Comment by SuperKogito** [comment]
Well, I already mentioned that I used a linter last time (autopep8 and yapf). So if I get it correctly, I should cancel the linter changes and take into account your new remarks. Or is it enough to fix the sections you mentioned?

**Comment by lostanlen** [comment]
@SuperKogito I would say it's fine. We'll keep your linter changes: there aren't that many. Please focus on addressing my CR comments. In the future, though, we'd appreciate if you put linter changes in a separate commit (if not a separate PR altogether).

File: librosa/feature/inverse.py (hunk 6 above)
Comment (SuperKogito) [reply librosa/feature/inverse.py]: This is not my favorite line and I agree with your concern. I replaced it by the following `np.where(mfcc == 0, np.finfo(float).eps, mfcc) ` which should help avoid the zero by zero division errors.

Comment (SuperKogito) [reply librosa/feature/inverse.py]: This change resulted in a failed build :( any suggestions on how to fix this?

Comment (SuperKogito) [reply librosa/feature/inverse.py]: I fixed it.

**Comment by bmcfee** [comment]
Thanks again to @SuperKogito and @lostanlen for the work on this PR/CR. We've all been busy for the past couple of weeks, but this seems pretty close to finished and I'd hate to see it languish from lack of attention.

@SuperKogito is there anything else you need from us to finish up the requested changes?

**Comment by SuperKogito** [comment]
@bmcfee From my side, I think this is done. However, I was waiting for the feedback from @lostanlen and you, especially about the last changes I implemented to handle division by zero and the FloatingPointError. If you are okay with those, then this PR is ready for merging.

File: librosa/feature/inverse.py [hunk 8]
```diff
@@ -195,26 +196,44 @@ def mfcc_to_mel(mfcc, n_mels=128, dct_type=2, norm='ortho', ref=1.0):
...
     scipy.fftpack.dct
     '''
+    if lifter > 0:
+        n_mfcc = mfcc.shape[0]
+        idx = np.arange(1, 1 + n_mfcc, dtype=mfcc.dtype)
+        lifter_sine = 1 + (lifter / 2) * np.sin(np.pi * idx / lifter)[:, np.newaxis]
+        lifter_sine = np.where(lifter_sine == 0, np.finfo(float).eps, lifter_sine)
+        mfcc = np.where(mfcc == 0, np.finfo(float).eps, mfcc) 
+        liftered_mfcc = mfcc / lifter_sine
+        mfcc = liftered_mfcc
```
Comment (bmcfee) [inline librosa/feature/inverse.py]: This part has a few issues that need to be fixed before we can merge.

1. Line 224: lifter_sine has exact zeros replaced by float epsilons.  Two things here: we'll almost never get exact zeros, and if you are going to replace these by machine epsilons, it should match the data type of `lifter_sine`, not be the default `float`.  You can get this in short-hand by calling `librosa.util.tiny(lifter_sine)`, but I'm not convinced this is a safe operation in general.
2. Why the in-filling of `mfcc` at line 225?  Dividing `0/lifter_sine` ought to be safe, but in general the `0/0` problem is undefined here, but I think any infilling should only be applied if we have a lifter that could zero out values.  If `lifter_sine` is always sufficiently positive, then we can safely infer that `mfcc==0` implies that the 0 value came pre-liftering, so post-liftering should stay 0 rather than becoming `eps/eps ~= 1`.  The same critique as above applies though, exact checks for 0 should be avoided.

Overall, I'm wary of trying too hard to prevent errors here.  Instead, I think it's safer to do the following:

1. Build the lifter array.
2. Check the lifter array for values likely to invoke underflow, and warn if they occur.
3. `mfcc = mfcc / lifter_sine` without any additional pre- or post-processing of `mfcc`.

**Comment by bmcfee** [comment]
> However, I was waiting for the feedback from @lostanlen and you, especially about the last changes I implemented to handle division by zero and the FloatingPointError.

Ah, sorry, github was doing that thing again where it's not giving the most recent view of the PR state.  I went in and added my thoughts to the CR, let me know what you think.

File: librosa/feature/inverse.py [hunk 9]
```diff
@@ -195,26 +197,47 @@ def mfcc_to_mel(mfcc, n_mels=128, dct_type=2, norm='ortho', ref=1.0):
...
     See Also
     --------
     mfcc
     melspectrogram
     scipy.fftpack.dct
     '''
+    if lifter > 0:
+        n_mfcc = mfcc.shape[0]
+        idx = np.arange(1, 1 + n_mfcc, dtype=mfcc.dtype)
+        lifter_sine = 1 + (lifter / 2) * np.sin(np.pi * idx / lifter)[:, np.newaxis]
```
Comment (bmcfee) [inline librosa/feature/inverse.py]: This will break in python 2.7 because it will use integer division instead of float.  You can fix this by saying `lifter * 0.5` instead of dividing by 2.

File: librosa/feature/inverse.py [hunk 10]
```diff
@@ -195,26 +197,47 @@ def mfcc_to_mel(mfcc, n_mels=128, dct_type=2, norm='ortho', ref=1.0):
...
 
-    logmel = scipy.fftpack.idct(mfcc, axis=0, type=dct_type, norm=norm, n=n_mels)
+        # check lifter array for critical values
+        if (lifter_sine == 0).any():
```
Comment (bmcfee) [inline librosa/feature/inverse.py]: Can we soften this to `np.abs(lifter_sine) < np.finfo(lifter_sine.dtype).eps`?  That way we can detect unstable points that aren't exactly on 0.

File: librosa/feature/inverse.py [hunk 11]
```diff
@@ -195,26 +197,47 @@ def mfcc_to_mel(mfcc, n_mels=128, dct_type=2, norm='ortho', ref=1.0):
...
+            warnings.warn("Warning: lifter array includes critial values that will likely to invoke underflow.")
```
Comment (bmcfee) [inline librosa/feature/inverse.py]: This will default to type `UserWarning` but the doc says `RuntimeWarning`.  I don't have a strong opinion about which is appropriate here, but it should be consistent.

**Comment by SuperKogito** [comment]
I am sorry that this PR is getting quite messy and ugly..I am facing some trouble running tests offline as I mentioned in  #1026. I agree with all your points and I am working on fixing the code but I keep making these stupid small mistakes (pushes and new PRs) every time I try to test or work around #1026.

**Comment by bmcfee** [comment]
> .I am facing some trouble running tests offline as I mentioned in #1026.

Is this still causing troubles for you?  Is there anything else we can do to help wrap this one up?

**Comment by bmcfee** [comment]
Your PR is showing a merge conflict because your editor has reformatted a bunch of unrelated code.  Could you revert those changes?  Thanks!

**Comment by SuperKogito** [comment]
@bmcfee I am working on it as we speak. I will make sure to reverse the formatting back, but first I need to fix the test that breaks with a `FloatingPointError: divide by zero encountered in true_divide`. I just managed to finally replicate the travis testing environment and its errors on my pc.  That's why it took me so long to respond.

However, I have a question concerning the raised `UserWarning` when this case occurs.  How should I go about that? right now I am testing by catching the warning but right after the test breaks due to the `FloatingPointError`, should I catch the error too or mark the test as failed to pass? is there a better option . You can take a look at my error logs [here](https://travis-ci.org/SuperKogito/librosa/jobs/630039825).

**Comment by bmcfee** [comment]
> However, I have a question concerning the raised `UserWarning` when this case occurs. How should I go about that? right now I am testing by catching the warning but right after the test breaks due to the `FloatingPointError`, should I catch the error too or mark the test as failed to pass?

I think it should be safe to divide by `lifter_sine + util.tiny(mfcc)` instead of `lifter_sine`.  If `lifter_sine` is near 0, then the corresponding mfcc dimension should also be, so you'd get `0 / (0 + eps) -> 0`.  If the mfcc dimension is not near 0, then it'll blow up, but that's why we have the warning.  If `lifter_sine` is far from 0, then this should have no measurable effect.

**Comment by SuperKogito** [comment]
Using `util.tiny()` is a great suggestion, and it seems to work like a charm. I tried to reverse the linter effects as much as I can but I am not sure how successful I was.

I thought it would be good to provide some feedback concerning my experience to help improve future contributor experiences: 
- To me the code was very readable & understandable and the communication with you guys and your guidance was great (really impressed by that and the welcoming culture and patience that you displayed) but replicating the development environment was very tricky especially since I worked on multiple projects on parallel and I used different devices to contribute. This might have been in part my fault and I might have not made the best choices when it comes to updating packages and my env but I still think that the `pytest` version requirement should be highlighted. Overall, it would be better if one can just get an anaconda development environment to be set-up automatically. In my case, every time I fixed a dependency something else was broken, that is why I did most of my last changes online and therefore I had to make a lot  of commits in comparison to the quantity of changes.
- I coded online and mostly relied on the Travis tests to check my code. However, in some cases some snippets did not display the same behaviour. For example: it took me a while to replicate the `FloatingPointError` on my machine. Instead I was getting a `RuntimeWarning` but using `np.seterr(divide='raise')` I got the error to show up. This might be useful for others to know.
- I also found the Travis tests kinda slow, therefore I think it is worthy taking a look at **distributed testing** using `pytest-xdist`.

**Comment by bmcfee** [comment]
> Using `util.tiny()` is a great suggestion, and it seems to work like a charm. I tried to reverse the linter effects as much as I can but I am not sure how successful I was.

Thanks, this all looks great to me!  There are a few dangling linter side effects, but I can live with that.  We do a pre-release audit for this kind of thing anyway, and it doesn't affect correctness.



> I thought it would be good to provide some feedback concerning my experience to help improve future contributor experiences:

Thanks for the feedback!  I'm glad we got this worked through, and appreciate your comments.  Some responses inline:



>     * but replicating the development environment was very tricky especially since I worked on multiple projects on parallel and I used different devices to contribute. This might have been in part my fault and I might have not made the best choices when it comes to updating packages and my env but I still think that the `pytest` version requirement should be highlighted.

Yeah, I don't have a great solution for this in general.  We do have all the requirements specified in the setup.py so that `pip install -e .[tests]` should do the trick (as stated in the contributing docs), but if there's anything that could be improved there, I'd be happy to revise things.



>     * I coded online and mostly relied on the Travis tests to check my code. However, in some cases some snippets did not display the same behaviour. For example: it took me a while to replicate the `FloatingPointError` on my machine. Instead I was getting a `RuntimeWarning` but using `np.seterr(divide='raise')` I got the error to show up. This might be useful for others to know.

This is kind of the point of having CI running on multiple versions / environments, so I take this as a success. :grin: 

It is problematic to *only* get testing through CI though, since it slows down progress substantially.  Again, if there's anything that could be made clearer in `CONTRIBUTING.md`, I think that would be the best place to clarify things.


>     * I also found the Travis tests kinda slow,

Agreed, but I don't think this is Travis's fault.  (Unlike AppVeyor, travis builds get scheduled and run pretty quickly.)

We have a long-standing issue to rewrite the test suite using more modern pytest (>4).  One side-effect of this rewrite would be better (re)use of test fixtures, which ought to bring down the running time of tests substantially.  It's currently at around 15min per build, and I'd like to get that down to under 5 if possible.

This is a huge undertaking though, and probably will take a couple of solid weeks of work by someone who knows the test suite well. (IE, probably me, though I'd be extremely happy for someone else to do it!)

It's also a little tangled up in #952, which if that ends up including some very short excerpts, could be helpful for accelerating certain parts of the test suite.

**Comment by bmcfee** [comment]
At any rate -- this PR looks great now, thanks for all the effort!  I'm happy to merge if @lostanlen wants to give the :+1:

**Comment by bmcfee** [comment]
:shipit: thanks again!
